pallets/crowdloans/src/lib.rs

```diff
@@ -878,6 +897,124 @@ pub mod pallet {
+            // 1. check origin (should be signed?)
+            let who = ensure_signed(origin)?;
+
+            log::trace!(
+                target: "crowdloans::claim_ptoken_share",
+                "crowdloan_id: {:?}, proof_block_hash: {:?}",
+                &crowdloan_id,
+                &proof_block_hash,
+            );
+
+            // 2. fetch crowdloan information
+            let vault = Self::current_vault(crowdloan_id).ok_or(Error::<T>::VaultDoesNotExist)?;
+
+            // 3. get derived token total issuance
+            let total_issuance = T::Assets::total_issuance(vault.ctoken);
+
+            // // 4. validate proof and extract useful information
+            let distribution = Self::distributions((crowdloan_id, proof_block_hash.encode()))
+                .ok_or(Error::<T>::DistributionDoesNotExist)?;
+
+            let storage_proof_and_blockhash_root = distribution.storage_hash;
+
+            // // convert trie path into proof
+            let storage_proof = StorageProof::new(proof.to_vec());
+
+            let local_result = sp_state_machine::read_proof_check::<BlakeTwo256, _>(
+                sp_core::hash::H256::from_slice(&storage_proof_and_blockhash_root),
+                storage_proof.clone(),
+                [proof_key],
+            )
+            .unwrap(); // _or(Error::<T>::InvaildProof)?;
+
+            // flatten bytes into varible for decoding
+            let local_result_as_bytes = local_result.into_iter().collect::<Vec<_>>();
+            let value_bytes = local_result_as_bytes[0].1.clone().unwrap();
+
+            // first 16 bytes are a u128 encoded as le
+            // this represents the balance
+            let mut balance_buffer: [u8; 16] = Default::default();
+            balance_buffer.copy_from_slice(&value_bytes[0..16]);
+
+            // convert to usable balance
+            let asset_account_balance = u128::from_le_bytes(balance_buffer);
+
+            // 5. check that proof matches distribution time
+
+            // throw error if incorrect blockhash
+            ensure!(
+                proof_block_hash == sp_core::hash::H256::from_slice(&distribution.block_hash),
+                Error::<T>::InvaildProof
+            );
+
+            // 6. get user proportional share
+            let share_of_ptokens =
+                sp_runtime::Perbill::from_rational(asset_account_balance, total_issuance);
+
+            // 7. calculate nominal amount of tokens to send user
+            // this will round down the to the hundredth
+            let nominal_amount =
+                (share_of_ptokens * distribution.amount).checked_div(1_000_000_000);
+
+            // TODO remove print statements
+            // println!("asset_account_balance\t{:#?}", asset_account_balance);
+            // println!("total_issuance\t{:#?}", total_issuance);
+            // println!("total_issuance\t{:#?}", share_of_ptokens);
+            // println!("nominal_amount\t{:#?}", nominal_amount);
```

Comment:
we can use substrate's built in logging and debugging lib too, that way it can all be controlled by a command line flag 😏 

See https://docs.substrate.io/v3/runtime/debugging/